### PR TITLE
Added bento charts library to replace recharts

### PR DESCRIPTION
**Pull request by SanjeevLakhwani**

### CODE REVIEW

**Review by davidlougheed — CHANGES_REQUESTED** [review]

File: package.json [hunk 1]
```diff
@@ -16,16 +17,15 @@
     "lodash.debounce": "^4.0.8",
     "prop-types": "^15.7.2",
     "query-string": "^6.13.7",
-    "react": "^16.14.0",
-    "react-dom": "^16.14.0",
+    "react": "^18.2.0",
```
[the remark this answers is not in the document]
Comment (SanjeevLakhwani) [reply package.json]: yes, anyways the PR was failing. should be resolved now.

File: src/components/charts/CustomPieChart.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    data,
+    chartHeight,
+    setAutoQueryPageTransition,
+    autoQueryDataType,
+    fieldLabel,
+    sort = true,
+}) => {
+    const history = useHistory();
+
+    const onClick = (data) => {
```
Comment (davidlougheed) [inline src/components/charts/CustomPieChart.js:25]: useCallback

File: src/components/charts/CustomPieChart.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        if (!setAutoQueryPageTransition || data.skipAutoquery) {
+            return;
+        }
+        console.log("data", data);
```
Comment (davidlougheed) [inline src/components/charts/CustomPieChart.js:29]: remove

File: src/components/charts/Histogram.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+const titleStyle = {
+    fontStyle: "italic",
+    padding: "0",
+    marginBottom: "-15px",
+};
+const titleHeaderHeight = 31;
+
+
+const Histogram = ({ title, data, chartHeight, chartAspectRatio }) => {
+    data = data.map(({ ageBin, count }) => ({ x: ageBin, y: count }));
```
Comment (davidlougheed) [inline src/components/charts/Histogram.js:15]: avoid mutating the param, instead assign this to a new const with a useMemo

File: src/template.html [hunk 5]
```diff
@@ -38,4 +38,4 @@
 <body>
     <div id="root"></div>
 </body>
-</html>
+</html>
```
Comment (davidlougheed) [inline src/template.html:41]: revert this change

**Review by davidlougheed — CHANGES_REQUESTED** [review]

**Comment by davidlougheed** [comment]
fix lint - prettier is conflicting with eslint here, so you need to run prettier and subsequently `npm run lint-fix`

File: src/template.html [hunk 6]
```diff
@@ -1,41 +1,41 @@
...
-        let BENTO_WEB_CONFIG = {};
-    </script>
-    <script src="<%= webpackConfig.output.publicPath %>static/config.js"></script>
-    <title><%= htmlWebpackPlugin.options.title %></title>
-    <style>
-        /* Special Override styles */
-        .ant-table .table-nested-ant-descriptions table {
-            border-collapse: collapse;
-        }
+    <head>
```
Comment (davidlougheed) [inline src/template.html:3]: needless change (indentation) again. please just do not change this file.

**Review by davidlougheed — CHANGES_REQUESTED** [review]

File: src/template.html (hunk 5 above)
Comment (davidlougheed) [reply src/template.html:39]: missing a terminal newline again

File: src/template.html [hunk 7]
```diff
@@ -19,7 +19,6 @@
         .ant-table .table-nested-ant-descriptions table {
             border-collapse: collapse;
         }
-
```
Comment (davidlougheed) [inline src/template.html:22]: no changes in template.html!

**Review by davidlougheed — CHANGES_REQUESTED** [review]

File: src/components/charts/ChartContainer.jsx [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+};
+
+const ChartContainer = ({ title, children }) => (
+    <div style={{ marginBottom: "20px", width: "420px" }}>
+        <h2 style={TITLE_STYLE}>{title}</h2>
+        {children}
+    </div>
+);
+
+export default ChartContainer;
```
Comment (davidlougheed) [inline src/components/charts/ChartContainer.jsx:15]: missing terminal newline

File: src/components/charts/CustomPieChart.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        (pointData) => {
+            if (!onAutoQueryTransition || pointData.skipAutoquery) return;
+
+            onAutoQueryTransition(window.location.href, dataType, labelKey, pointData.name);
+            history.push("/data/explorer/search");
+        },
+        [onAutoQueryTransition, dataType, labelKey, history]
+    );
+
+    if (!data.length) {
```
Comment (v-rocheleau) [inline src/components/charts/CustomPieChart.js:29]: Will error if data prop is undefined or null. A safer check in JS could be:
```

**Review by davidlougheed — CHANGES_REQUESTED** [review]
other than minor thing and victors comment, lgtm

**Review by v-rocheleau — CHANGES_REQUESTED** [review]
Small changes. Will need a lint fix.

File: src/components/charts/Histogram.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+const transformData = (data) =>
+    data.map(({ ageBin, count }) => ({ x: ageBin, y: count }));
+
+const Histogram = ({
+    title = "Histogram",
+    data = [],
+    chartHeight = 300,
+}) => {
+
+    if (!data || !data.length) {
```
[the remark this answers is not in the document]
Comment (SanjeevLakhwani) [reply src/components/charts/Histogram.js:16]: I get it but I cannot be not an array/null/undefined cause of proptype check

File: src/components/charts/Histogram.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
+import React, { useMemo } from "react";
+import PropTypes from "prop-types";
+import { Empty } from "antd";
+import { BarChart } from "bento-charts";
+import ChartContainer from "./ChartContainer";
+
+const transformData = (data) =>
+    data.map(({ ageBin, count }) => ({ x: ageBin, y: count }));
+
+const Histogram = ({
```
[the remark this answers is not in the document]
Comment (SanjeevLakhwani) [reply src/components/charts/Histogram.js:10]: Good point, would change.

File: src/components/charts/PieChart.js [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        return (
+            <ChartContainer title={title}>
+                <div style={{ height: chartHeight, display: "flex", alignItems: "center", justifyContent: "center" }}>
+                    <Empty image={Empty.PRESENTED_IMAGE_SIMPLE} description="No available data" />
+                </div>
+            </ChartContainer>
+        );
+    }
+
+    const pieChartData = data.map(({ name, value }) => ({ x: name, y: value }));
```
Comment (davidlougheed) [inline src/components/charts/PieChart.js:39]: memoize either component or this

**Review by davidlougheed — CHANGES_REQUESTED** [review]

File: src/components/charts/PieChart.js [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+        return (
+            <ChartContainer title={title}>
+                <div style={{ height: chartHeight, display: "flex", alignItems: "center", justifyContent: "center" }}>
+                    <Empty image={Empty.PRESENTED_IMAGE_SIMPLE} description="No available data" />
+                </div>
+            </ChartContainer>
+        );
+    }
+
+    const pieChartData = useMemo(data.map(({ name, value }) => ({ x: name, y: value })), data);
```
Comment (davidlougheed) [inline src/components/charts/PieChart.js:39]: bad syntax for usememo

File: src/components/charts/PieChart.js [hunk 14]
```diff
@@ -0,0 +1,63 @@
+import React, {useCallback, useMemo} from "react";
```
Comment (davidlougheed) [inline src/components/charts/PieChart.js:1]: use prettier spacing here since youve applied prettier everywhere else

**Review by davidlougheed — CHANGES_REQUESTED** [review]

**Comment by davidlougheed** [comment]
lint please

File: src/components/charts/Histogram.js [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+            <BarChart data={transformedData} height={chartHeight} units="" />
+        </ChartContainer>
+    );
+};
+
+Histogram.propTypes = {
+    title: PropTypes.string,
+    data: PropTypes.arrayOf(
+        PropTypes.shape({
+            ageBin: PropTypes.string.isRequired,
```
Comment (v-rocheleau) [inline src/components/charts/Histogram.js:22]: We may want to display histogram data that is not related to patients age, so I would make the props more generic.
Since `transformData` only maps `ageBin` to `x` and `count` to `y` I think we can simply use those as our prop types and get rid of `transformData`.

**Review by v-rocheleau — CHANGES_REQUESTED** [review]
props tweaks

**Review by davidlougheed — APPROVED** [review]

**Review by v-rocheleau — APPROVED** [review]
